### DISCUSSION TITLE
Extension of `user` in `AuthSession`

### DISCUSSION THREAD

**Discussion by seblyng** [maxcountryman/axum-login#139] [Q&A] [answered]
Hi :) Sorry if this is a stupid question and it is already possible. But previously, I was able to do something like this:

```rs
pub fn home(Extension(user): Extension<user>) -> impl IntoResponse {
    todo!();
}
```

And I would have access to the user immediately.

After the updates lately, I can only see docs and examples for getting it from `AuthSession`, so I have to do something like this:

```rs
pub fn home(auth: AuthSession) -> impl IntoResponse {
    let user = auth.user;
    todo!();
}
```

I really liked to be able to get the user with an extension like I could previously. Is this possible in some way?

**Comment by maxcountryman** [comment, marked as answer]
We removed the `User` extension because it seemed redundant with the session.

If you feel like it simplifies your code, you could write something like this instead:

```rust
pub fn home(AuthSession { user, ... }: AuthSession) -> impl IntoResponse {
    todo!();
}
```

I don't think we're opposed to adding the extension again, but it's not clear it's really needed.

**Reply by seblyng** [reply]
Ok, I thought about it some more right after I asked the question, and I think I agree with you that it is not needed!

Thank you!

**Comment by arpankapoor** [comment]
If I'm not wrong, the `Extension` gave direct access to the `User` instead of a `Option<User>` as is the case with `AuthSession`. Handling the `Option` inside a protected route feels a bit redudant IMHO. So adding back the extension would be a welcome addition!

Thanks for all your work on this crate!

**Reply by maxcountryman** [reply]
Something like this is possible:

```diff
diff --git a/axum-login/src/service.rs b/axum-login/src/service.rs
index f25e85d..7cdfce1 100644
--- a/axum-login/src/service.rs
+++ b/axum-login/src/service.rs
@@ -86,7 +86,11 @@ where
                     }
                 };

-                req.extensions_mut().insert(auth_session);
+                req.extensions_mut().insert(auth_session.clone());
+
+                if let Some(user) = auth_session.user {
+                    req.extensions_mut().insert(user);
+                }

                 inner.call(req).await
             }
```

However, handlers would need to use the `Extension` extractor to retrieve the user (and notice this also introduces a clone of the `auth_session` in the service).
